.github/workflows/runtests3.py

```diff
@@ -0,0 +1,120 @@
+import os
+import subprocess
+import shlex
+import tempfile
+import threading
+import ssl
+import shutil
+from http.server import HTTPServer, SimpleHTTPRequestHandler
+
+def run_command(cmd, expected_retcode=0):
+    """Runs the supplied command and checks for the expected return code."""
+    retcode = subprocess.call(shlex.split(cmd))
+    if retcode != expected_retcode:
+        raise Exception(f"Expected {expected_retcode} from process but it exited with {retcode}.")
+
+# Cleanup previous runs
+TEST_CLONE_DIR = "test_clone"
+for path in [TEST_CLONE_DIR, 'repo']:
+    if os.path.exists(path):
+        shutil.rmtree(path)
+        print(f"Cleaned up existing directory: {path}")
+
+# Create temporary server environment
+with tempfile.TemporaryDirectory() as server_dir, tempfile.TemporaryDirectory() as local_repo_dir:
+    # Setup bare repository at correct path
+    bare_repo_path = os.path.join(server_dir, 'repo.git')
+    subprocess.run(['git', 'init', '--bare', bare_repo_path], check=True)
+    
+    # Enable HTTP transport support
+    subprocess.run(['git', '-C', bare_repo_path, 'config', 'http.receivepack', 'true'], check=True)
+    subprocess.run(['git', '-C', bare_repo_path, 'config', 'http.uploadpack', 'true'], check=True)
+    
+    # Initialize test repository
+    subprocess.run(['git', 'init', local_repo_dir], check=True)
+    with open(os.path.join(local_repo_dir, 'README.md'), 'w') as f:
+        f.write('# Local Test Repository\n')
+    subprocess.run(['git', '-C', local_repo_dir, 'add', '.'], check=True)
+    subprocess.run(['git', 'config', '--global', 'user.email', "'you@example.com'"],check =True)
+    subprocess.run(['git', 'config', '--global', 'user.name', "'Your Name'"],check =True)
```

Comment:
Let's make these local configuration options in case we need to try this out on a machine. This will mean the other repo setup commands need to be updated to include these as well.
```suggestion
    subprocess.run(['git', 'config', '-C', local_repo_dir, '--local', 'user.email', "'you@example.com'"],check =True)
    subprocess.run(['git', 'config', '-C', local_repo_dir, '--local', 'user.name', "'Your Name'"],check =True)
```